FuelStop doc (#1)
circle-free/fuelstop-examples: README.md

# FuelStop

## Getting Started

#### Projects

> At FuelStop, you should have one project for each Ethereum network. After sign up, you can find your projects by clicking on the menu icon on the top-left corner. By default, one should have one project Mainnet and one project for Ropsten (which is a proof-of-work Ethereum testnet) after sign up. Clicking on any of the project will get you to the project details page.

Reviewer: ```suggestion
After sign up, you will have 2 projects, one for each Ethereum network.  You can find your projects by clicking on the menu icon on the top-left corner on the page. One project is for Mainnet and the other is for Ropsten (a proof-of-work Ethereum testnet). Clicking on any of the projects will bring you to it's project details page.
```

![menu](./diagrams/Projects.svg)

#### Project Details

> The project details page shows your balance, your unique JSON-RPC url that you should use that contains your `Project ID` to authenticate the API calls, your registered addresses, your transactions history and account history.

Reviewer: ```suggestion
The project details page includes the project's balance, registered addresses, transaction history, account history, and unique JSON-RPC url containing the Project ID which is used to authenticate your API calls.
```

###### Balance

This section is where you can add your credit card with FuelStop, and after which displays information about the account associated with the project. The Mainnet project would start without credit card added, and therefore you have to add one before we can fund your transactions. The Ropsten project, however, is with a testnet of Ethereum, and does not require you to add any credit card to start using. It comes with a fake credit card, and you can top it up anytime you want by clicking on the "Top Up Button" with no cost. This is where you can test your system or library with FuelStop.

![Balance without credit card](./diagrams/balance-without-credit-card.svg)

![FuelStop Documentation](./diagrams/balance.svg)

###### Details

> Here you can find the name of the network the project is associated with, and your RPC URL. FuelStop's Ethereum JSON-RPC APIs require a valid project ID to be part of your URL to autheticate the calls. This is how we identify who you are behind the API requests, and it is important to keep it as secret to yourself.

Reviewer: ```suggestion
Here you can find the name of the network the project is associated with, and your RPC URL. FuelStop's Ethereum JSON-RPC APIs require a valid project ID to be part of your URL to authenticate the calls. This is how we identify who you are behind the API requests, and it is important to keep it a secret.
```

![Details](./diagrams/details.svg)

###### Addresses

> Before sending your signed transactions to us via `eth_sendRawTransaction`, you have to register your address with us. If a signed transaction that is not from any of the registered addresses gets sent to FuelStop via your RPC URL, FuelStop will behave no differently than regular Ethereum nodes and simply broadcase it the network. If your address has some Ether but not enough to cover the particular transaction's fees, FuelStop would only fund your address to the  exact amount that is enough for the transaction to go through.

Reviewer: ```suggestion
Before sending your signed transactions to us via `eth_sendRawTransaction`, you need to register your address with us. If you send FuelStop a signed transaction via your RPC URL that is not included in your registered addresses, FuelStop will behave no differently than a regular Ethereum node and will broadcast it to the network. If your address has some ETH but not enough to cover the transaction's fees, FuelStop will fund your address with the exact amount needed for the transaction to go through.
```

![Addresses](./diagrams/addresses.svg)

###### Transactions

> Transactions table contains your transaction history. You can keep refreshing to monitor the status of your transactions. Your transaction status could be as follows.

Reviewer: ```suggestion
The transactions table contains your transaction history. You can periodically refresh it to monitor the status of your transactions. A transaction's status will be one of the following:
```

1. `New`: The transaction has been received and is waiting for the fund transaction to be mined.

> 2. `Sent`: The transaction has been broadcast to the network and is waiting for it to be mined.

Reviewer: ```suggestion
2. `Sent`: The transaction has been broadcasted to the network and is waiting to be mined.
```

3. `Confirmed`: The transaction has been mined in a block and transaction execution is successful on-chain.

> 4. `Failed`: The transaction has been mined in a block and transaction execution is failed on-chain. This is likely because you sent your transaction to a smart contract, and part of the contract logic decided to reject any state transition.

Reviewer: ```suggestion
4. `Failed`: The transaction has been mined in a block and transaction execution has failed on-chain. This is likely because you sent your transaction to a smart contract, and part of the contract logic has rejected a state transition.
```

![Transactions](./diagrams/transactions.svg)

###### Account History

> Account History table displays a history of your account balance change.

Reviewer: ```suggestion
The Account History table displays your account balance change history.
```

![Account History](./diagrams/account-history.svg)

#### Using FuelStop's RPC URL

FuelStop supports all [standard Ethereum JSON-RPC APIs](https://eth.wiki/json-rpc/API), which means it is able to work seemlessly with popular libraries such as [web3.js](https://github.com/ethereum/web3.js/) and [ethers.js](https://github.com/ethers-io/ethers.js/), and tools such as [Metamask](https://metamask.io/) and [Remix Ethereum IDE](https://remix.ethereum.org/).

###### web3.js

```javascript
const Web3 = require('web3');
const web3 = new Web3(url); // url is your FuelStop project RPC URL
```

###### ethers.js

```javascript
const ethers = require('ethers');
const provider = new ethers.providers.JsonRpcProvider(url); // url is your FuelStop project RPC URL
```

###### Metamask

> Go to `Settings` -> `Networks` -> `Add Network`, and then fill in the "Network Name" of your choice and your FuelStop RPC URL, and your Metamask is connected.

Reviewer: ```suggestion
To connect Fuelstop to Metamask
1. Go to `Settings` -> `Networks` -> `Add Network`
2. Fill in the "Network Name" of your choice
3. Add your FuelStop RPC URL
```

![Metamask](./diagrams/metamask.svg)

> If you have loaded the private key in Metamask of some addresses that you also registered in FuelStop, you may notice that it's ETH balacne shown in Metamask does not equal to its actual balance on the Ethereum network. This is because FuelStop intercepts the RPC calls that Metamask uses to query your balance when it sees those calls are for an address registered for the corresponding project. The balance after interception is your `actual ETH balance` + `equivalent ETH amount of FuelStop project balance`. This effectively would allow Metamask to work with FuelStop seemlessly when it comes to sending transactions, and is useful for you to continue using Metamask to monitor your "ETH" balance.

Reviewer: ```suggestion
If you register a address to Fuelstop that is associated with one of your Metmask accounts, you may notice that that the ETH balance shown in Metamask is not equal to its actual balance on the Ethereum network. This is because FuelStop intercepts the RPC calls that Metamask uses to query your balance. When FuelStop sees a call querying for a balance of a registered address it will give back your `actual ETH balance` + `the equivalent ETH amount of your FuelStop project balance`. This allows Metamask to work with FuelStop seamlessly when it comes to sending transactions, and is useful for you to continue using Metamask to monitor your "ETH" balance.
```

![Metamask Balance](./diagrams/metamask-balance.svg)

###### Remix Ethereum IDE

> To connect the Remix Ethereum IDE with FuelStop, you will need to either follow the Metamask section to connect it with Metamask first and then select "Injected Web3", or you can select "Web3 Provider" and just enter your FuelStop RPC URL.

Reviewer: ```suggestion
To connect the Remix Ethereum IDE with FuelStop, you will need to either follow the Metamask section to connect it with Metamask first and then select "Injected Web3" in the Remix IDE, or you can select "Web3 Provider" and enter your FuelStop RPC URL.
```

# FuelStop

## Getting Started

> #### Projects

Reviewer 2: ```suggestion
### Projects
```

> After sign up, you will have 2 projects, one for each Ethereum network.  You can find your projects by clicking on the menu icon on the top-left corner on the page. One project is for Mainnet and the other is for Ropsten (a proof-of-work Ethereum testnet). Clicking on any of the projects will bring you to it's project details page.

Reviewer 2: ```suggestion
After sign up, you can find your projects by clicking on the menu icon on the top-left corner of the page. One project is for the Ethereum Mainnet and the other is for Ropsten (this is the only testnet that's currently supported). Clicking on either of the projects will bring you to its details page.
```

![menu](./diagrams/projects.svg)

> #### Project Details

Reviewer 2: ```suggestion
### Project Details
```

> The project details page includes the project's balance, registered addresses, transaction history, account history, and unique JSON-RPC url containing the Project ID which is used to authenticate your API calls.

Reviewer 2: ```suggestion
The `Project Details` page includes the project's balance, registered addresses, transaction history, account history, and the unique JSON-RPC URL containing the Project ID which is used to authenticate your API calls.
```

###### Balance

> This section is where you can add your credit card with FuelStop, and after which displays information about the account associated with the project. The Mainnet project would start without credit card added, and therefore you have to add one before we can fund your transactions. The Ropsten project, however, is with a testnet of Ethereum, and does not require you to add any credit card to start using. It comes with a fake credit card, and you can top it up anytime you want by clicking on the "Top Up Button" with no cost. This is where you can test your system or library with FuelStop.

Reviewer 2: ```suggestion
The `Balance` section is where you can add your credit card to your FuelStop account. After the card is added, its info is displayed. You can also remove a card once it has been added. The `Mainnet` project starts without credit-card so, you have to add one before your transactions can be paid for. The `Ropsten` project, however, has a test-card already registered that cannot be removed. Feel free to hit `Top up` as much as you want with this project, you won't be charged.
```

![Balance without credit card](./diagrams/balance-without-credit-card.svg)

![FuelStop Documentation](./diagrams/balance.svg)

###### Details

> Here you can find the name of the network the project is associated with, and your RPC URL. FuelStop's Ethereum JSON-RPC APIs require a valid project ID to be part of your URL to authenticate the calls. This is how we identify who you are behind the API requests, and it is important to keep it a secret.

Reviewer 2: ```suggestion
Here, you can find the name of the network and your RPC URL. FuelStop's Ethereum JSON-RPC APIs require a valid project ID to be part of your URL to authenticate the calls. Do not share the URL.
```

> ###### Balance
> 
> The `Balance` section is where you can add your credit card to your FuelStop account. After the card is added, its info is displayed. You can also remove a card once it has been added. The `Mainnet` project starts without credit-card so, you have to add one before your transactions can be paid for. The `Ropsten` project, however, has a test-card already registered that cannot be removed. Feel free to hit `Top up` as much as you want with this project, you won't be charged.
> 
> ![Balance without credit card](./diagrams/balance-without-credit-card.svg)
> 
> ![FuelStop Documentation](./diagrams/balance.svg)
> 
> ###### Details
> 
> Here, you can find the name of the network and your RPC URL. FuelStop's Ethereum JSON-RPC APIs require a valid project ID to be part of your URL to authenticate the calls. Do not share the URL.
> 
> ![Details](./diagrams/details.svg)
> 
> ###### Addresses
> 
> Before sending your signed transactions (via the JSON-RPC `eth_sendRawTransaction` method or with a client-library), you need to register your addresses. If you send FuelStop a signed transaction that is not included in your address-list, FuelStop will behave no differently than a regular Ethereum node and will broadcast it to the network. If your address has some ETH but not enough to cover the transaction's fees, FuelStop will fund your address so that the transaction can go through.
> 
> ![Addresses](./diagrams/addresses.svg)
> 
> ###### Transactions

Reviewer 2: ```suggestion
#### Balance

The `Balance` section is where you can add your credit card to your FuelStop account. After the card is added, its info is displayed. You can also remove a card once it has been added. The `Mainnet` project starts without credit-card so, you have to add one before your transactions can be paid for. The `Ropsten` project, however, has a test-card already registered that cannot be removed. Feel free to hit `Top up` as much as you want with this project, you won't be charged.

![Balance without credit card](./diagrams/balance-without-credit-card.svg)

![FuelStop Documentation](./diagrams/balance.svg)

#### Details

Here, you can find the name of the network and your RPC URL. FuelStop's Ethereum JSON-RPC APIs require a valid project ID to be part of your URL to authenticate the calls. Do not share the URL.

![Details](./diagrams/details.svg)

#### Addresses

Before sending your signed transactions (via the JSON-RPC `eth_sendRawTransaction` method or with a client-library), you need to register your addresses. If you send FuelStop a signed transaction that is not included in your address-list, FuelStop will behave no differently than a regular Ethereum node and will broadcast it to the network. If your address has some ETH but not enough to cover the transaction's fees, FuelStop will fund your address so that the transaction can go through.

![Addresses](./diagrams/addresses.svg)

#### Transactions
```

> Before sending your signed transactions to us via `eth_sendRawTransaction`, you need to register your address with us. If you send FuelStop a signed transaction via your RPC URL that is not included in your registered addresses, FuelStop will behave no differently than a regular Ethereum node and will broadcast it to the network. If your address has some ETH but not enough to cover the transaction's fees, FuelStop will fund your address with the exact amount needed for the transaction to go through.

Reviewer 2: ```suggestion
Before sending your signed transactions (via the JSON-RPC `eth_sendRawTransaction` method or with a client-library), you need to register your addresses. If you send FuelStop a signed transaction that is not included in your address-list, FuelStop will behave no differently than a regular Ethereum node and will broadcast it to the network. If your address has some ETH but not enough to cover the transaction's fees, FuelStop will fund your address so that the transaction can go through.
```

> The transactions table contains your transaction history. You can periodically refresh it to monitor the status of your transactions. A transaction's status will be one of the following:

Reviewer 2: ```suggestion
The `Transactions` table contains your transaction history. You can periodically refresh it to monitor the status of your transactions. A transaction's status will be one of the following:
```

> 1. `New`: The transaction has been received and is waiting for the fund transaction to be mined.

Reviewer 2: ```suggestion
1. `Pending`: The transaction has been received and is queued to be sent out once the address has been funded.
```

Reviewer 2: You don't even know your own platform... `New`.... pfff

2. `Sent`: The transaction has been broadcasted to the network and is waiting to be mined.
3. `Confirmed`: The transaction has been mined in a block and transaction execution is successful on-chain.
4. `Failed`: The transaction has been mined in a block and transaction execution has failed on-chain. This is likely because you sent your transaction to a smart contract, and part of the contract logic has errored out.

![Transactions](./diagrams/transactions.svg)

> ###### Account History

Reviewer 2: ```suggestion
#### Account History
```

> 4. `Failed`: The transaction has been mined in a block and transaction execution has failed on-chain. This is likely because you sent your transaction to a smart contract, and part of the contract logic has rejected a state transition.

Reviewer 2: ```suggestion
4. `Failed`: The transaction has been mined in a block and transaction execution has failed on-chain. This is likely because you sent your transaction to a smart contract, and part of the contract logic has errored out.
```

Reviewer 2: I'm not sure if there's a better way to say it than "errored out", but we shouldn't mention a "state transition" as that's part of the inner workings of the network, and you don't need to understand that to be able to use our platform.

> The Account History table displays your account balance change history.

Reviewer 2: ```suggestion
The Account History table displays your account's balance-changes.
```

![Account History](./diagrams/account-history.svg)

#### Using FuelStop's RPC URL

> FuelStop supports all [standard Ethereum JSON-RPC APIs](https://eth.wiki/json-rpc/API), which means it is able to work seemlessly with popular libraries such as [web3.js](https://github.com/ethereum/web3.js/) and [ethers.js](https://github.com/ethers-io/ethers.js/), and tools such as [Metamask](https://metamask.io/) and [Remix Ethereum IDE](https://remix.ethereum.org/).

Reviewer 2: ```suggestion
FuelStop supports all [standard Ethereum JSON-RPC APIs](https://eth.wiki/json-rpc/API), which means it's able to work seemlessly with popular libraries such as [web3.js](https://github.com/ethereum/web3.js/) and [ethers.js](https://github.com/ethers-io/ethers.js/), and tools such as [MetaMask](https://metamask.io/) and [Remix Ethereum IDE](https://remix.ethereum.org/).
```

###### web3.js

```javascript
const Web3 = require('web3');
const web3 = new Web3(url); // url is your FuelStop project RPC URL
```

###### ethers.js

```javascript
const ethers = require('ethers');
const provider = new ethers.providers.JsonRpcProvider(url); // url is your FuelStop project RPC URL
```

> ###### Metamask

Reviewer 2: ```suggestion
###### MetaMask
```

To connect Fuelstop to Metamask
1. Go to `Settings` -> `Networks` -> `Add Network`
2. Fill in the "Network Name" of your choice
3. Add your FuelStop RPC URL

> ![Metamask](./diagrams/metamask.svg)

Reviewer 2: ```suggestion
![MetaMask](./diagrams/metamask.svg)
```

> #### Using FuelStop's RPC URL
> 
> FuelStop supports all [standard Ethereum JSON-RPC APIs](https://eth.wiki/json-rpc/API), which means it's able to work seemlessly with popular libraries such as [web3.js](https://github.com/ethereum/web3.js/) and [ethers.js](https://github.com/ethers-io/ethers.js/), and tools such as [MetaMask](https://metamask.io/) and [Remix Ethereum IDE](https://remix.ethereum.org/).
> 
> ###### web3.js
> 
> ```javascript
> const Web3 = require('web3');
> const web3 = new Web3(url); // url is your FuelStop project RPC URL
> ```
> 
> ###### ethers.js
> 
> ```javascript
> const ethers = require('ethers');
> const provider = new ethers.providers.JsonRpcProvider(url); // url is your FuelStop project RPC URL
> ```
> 
> ###### MetaMask
> 
> To connect Fuelstop to Metamask
> 1. Go to `Settings` -> `Networks` -> `Add Network`
> 2. Fill in the "Network Name" of your choice
> 3. Add your FuelStop RPC URL
> 
> ![MetaMask](./diagrams/metamask.svg)
> 
> If you register one of your MetaMask addresses with a Fuelstop project, you may notice that that the ETH balance shown in MetaMask is higher than the actual balance on the Ethereum network. Instead, the amount is the `actual ETH balance` + `the equivalent ETH amount of your FuelStop project balance`. This allows MetaMask to work seamlessly with FuelStop even when your MetaMask address has no balance.
> 
> ![Metamask Balance](./diagrams/metamask-balance.svg)
> 
> ###### Remix Ethereum IDE

Reviewer 2: ```suggestion
### Using FuelStop's RPC URL

FuelStop supports all [standard Ethereum JSON-RPC APIs](https://eth.wiki/json-rpc/API), which means it's able to work seemlessly with popular libraries such as [web3.js](https://github.com/ethereum/web3.js/) and [ethers.js](https://github.com/ethers-io/ethers.js/), and tools such as [MetaMask](https://metamask.io/) and [Remix Ethereum IDE](https://remix.ethereum.org/).

#### web3.js

```javascript
const Web3 = require('web3');
const web3 = new Web3(url); // url is your FuelStop project RPC URL
```

#### ethers.js

```javascript
const ethers = require('ethers');
const provider = new ethers.providers.JsonRpcProvider(url); // url is your FuelStop project RPC URL
```

#### MetaMask

To connect Fuelstop to Metamask
1. Go to `Settings` -> `Networks` -> `Add Network`
2. Fill in the "Network Name" of your choice
3. Add your FuelStop RPC URL

![MetaMask](./diagrams/metamask.svg)

If you register one of your MetaMask addresses with a Fuelstop project, you may notice that that the ETH balance shown in MetaMask is higher than the actual balance on the Ethereum network. Instead, the amount is the `actual ETH balance` + `the equivalent ETH amount of your FuelStop project balance`. This allows MetaMask to work seamlessly with FuelStop even when your MetaMask address has no balance.

![Metamask Balance](./diagrams/metamask-balance.svg)

#### Remix Ethereum IDE
```

> If you register a address to Fuelstop that is associated with one of your Metmask accounts, you may notice that that the ETH balance shown in Metamask is not equal to its actual balance on the Ethereum network. This is because FuelStop intercepts the RPC calls that Metamask uses to query your balance. When FuelStop sees a call querying for a balance of a registered address it will give back your `actual ETH balance` + `the equivalent ETH amount of your FuelStop project balance`. This allows Metamask to work with FuelStop seamlessly when it comes to sending transactions, and is useful for you to continue using Metamask to monitor your "ETH" balance.

Reviewer 2: ```suggestion
If you register one of your MetaMask addresses with a Fuelstop project, you may notice that that the ETH balance shown in MetaMask is higher than the actual balance on the Ethereum network. Instead, the amount is the `actual ETH balance` + `the equivalent ETH amount of your FuelStop project balance`. This allows MetaMask to work seamlessly with FuelStop even when your MetaMask address has no balance.
```

> To connect the Remix Ethereum IDE with FuelStop, you will need to either follow the Metamask section to connect it with Metamask first and then select "Injected Web3" in the Remix IDE, or you can select "Web3 Provider" and enter your FuelStop RPC URL.

Reviewer 2: ```suggestion
To connect FuelStop to the Remix Ethereum IDE, navigate the `DEPLOY & RUN TRANSACTIONS`. In the `ENVIRONMENT` dropdown, you can either select "Web3 Provider" and enter your Fuelstop RPC-URL **OR** you can connect FuelStop to MetaMask (following the section above) then select "Injected Web3" in the Remix IDE.
```
